File: backend/src/test/scala/kpn/api/custom/TagsTest.scala
package kpn.api.custom

import kpn.api.common.data.Tagable
import kpn.core.util.UnitTest

class TagsTest extends UnitTest {

  private case class TestObject(tags: Seq[Tag]) extends Tagable

  test("hasTag - false when tag is missing") {
    TestObject(Seq.empty).hasTag("key") should equal(false)
  }

  test("hasTag - tag with single value") {
    val taggable = taggableWithTagValue("value")
    taggable.hasTag("key") should equal(true)
    taggable.hasTag("key", "value") should equal(true)
    taggable.hasTag("key", "bla") should equal(false)
  }

  test("hasTag - tag with multiple values") {
    val taggable = taggableWithTagValue("value1;value2")
    taggable.hasTag("key", "value1") should equal(true)
    taggable.hasTag("key", "value2") should equal(true)
    taggable.hasTag("key", "bla") should equal(false)
  }

  test("values - no values when tag is missing") {
    TestObject(Seq.empty).tagValues("key") should equal(Seq.empty)
  }

  test("values - no values when tag value is empty") {
    taggableWithTagValue("").tagValues("key") should equal(Seq.empty)
  }

  test("values") {
    taggableWithTagValue("value").tagValues("key") should equal(Seq("value"))
    taggableWithTagValue("  value  ").tagValues("key") should equal(Seq("value"))
    taggableWithTagValue("value1;value2").tagValues("key") should equal(Seq("value1", "value2"))
    taggableWithTagValue("value1;;value2").tagValues("key") should equal(Seq("value1", "value2"))
    taggableWithTagValue("value1  ;  ;  value2").tagValues("key") should equal(Seq("value1", "value2"))
  }

  private def taggableWithTagValue(value: String): TestObject = {
    TestObject(Tags.from("key" -> value))
  }
}

File: backend/api/src/test/java/kpn/api/custom/TagsTest.java
package kpn.api.custom;

import kpn.api.common.data.Tagable;
import org.junit.jupiter.api.Test;

import com.google.common.collect.ImmutableList;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class TagsTest {

  private record TestObject(
    ImmutableList<Tag> tags
  ) implements Tagable {}

  @Test
  public void hasTag_falseWhenTagIsMissing() {
    TestObject taggable = new TestObject(ImmutableList.of());
    assertFalse(taggable.hasTag("key"));
  }

  @Test
  public void hasTag_tagWithSingleValue() {
    TestObject taggable = taggableWithTagValue("value");
    assertTrue(taggable.hasTag("key"));
    assertTrue(taggable.hasTag("key", "value"));
    assertFalse(taggable.hasTag("key", "bla"));
  }

  @Test
  public void hasTag_tagWithMultipleValues() {
    TestObject taggable = taggableWithTagValue("value1;value2");
    assertTrue(taggable.hasTag("key", "value1"));
    assertTrue(taggable.hasTag("key", "value2"));
    assertFalse(taggable.hasTag("key", "bla"));
  }

  @Test
  public void tagValues_noValuesWhenTagIsMissing() {
    TestObject taggable = new TestObject(ImmutableList.of());
    assertEquals(taggable.tagValues("key"), ImmutableList.of());
  }

  @Test
  public void values_noValuesWhenTagValueIsEmpty() {
    assertEquals(taggableWithTagValue("").tagValues("key"), ImmutableList.of());
  }

  @Test
  public void tagValues() {
    assertEquals(taggableWithTagValue("value").tagValues("key"), ImmutableList.of("value"));
    assertEquals(taggableWithTagValue("  value  ").tagValues("key"), ImmutableList.of("value"));
    assertEquals(taggableWithTagValue("value1;value2").tagValues("key"), ImmutableList.of("value1", "value2"));
    assertEquals(taggableWithTagValue("value1;;value2").tagValues("key"), ImmutableList.of("value1", "value2"));
    assertEquals(taggableWithTagValue("value1  ;  ;  value2").tagValues("key"), ImmutableList.of("value1", "value2"));
  }

  private static TestObject taggableWithTagValue(String value) {
    return new TestObject(
      ImmutableList.of(
        new Tag("key", value)
      )
    );
  }
}
